### PR TITLE
tune: selecting notebooks

### PR DESCRIPTION
**Pull request by andrewferlitsch**
**REQUIRED:** Add a summary of your PR here, typically including why the change is needed and what was changed. Include any design alternatives for discussion purposes.

<br>
Add selecting build results by date range
Tune the selection algorithm
Add initial test
<br><br><br>

**REQUIRED:** Fill out the below checklists or remove if irrelevant
1. If you are opening a PR for `Official Notebooks` under the [notebooks/official](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/notebooks/official) folder, follow this mandatory checklist:
- [ ] Use the [notebook template](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/notebook_template.ipynb) as a starting point.
- [ ] Follow the style and grammar rules outlined in the above notebook template.
- [ ] Verify the notebook runs successfully in Colab since the automated tests cannot guarantee this even when it passes.
- [ ] Passes all the required automated checks. You can locally test for formatting and linting with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).
- [ ] You have consulted with a tech writer to see if tech writer review is necessary. If so, the notebook has been reviewed by a tech writer, and they have approved it.
- [ ] This notebook has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/official/CODEOWNERS) file under the `Official Notebooks` section, pointing to the author or the author's team.
- [ ] The Jupyter notebook cleans up any artifacts it has created (datasets, ML models, endpoints, etc) so as not to eat up unnecessary resources.

<br>

2. If you are opening a PR for `Community Notebooks` under the [notebooks/community](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/notebooks/community) folder:
- [ ] This notebook has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/community/CODEOWNERS) file under the `Community Notebooks` section, pointing to the author or the author's team.
- [ ] Passes all the required formatting and linting checks. You can locally test with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).

<br>

3. If you are opening a PR for `Community Content` under the [community-content](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/community-content) folder:
- [ ] Make sure your main `Content Directory Name` is descriptive, informative, and includes some of the key products and attributes of your content, so that it is differentiable from other content
- [ ] The main content directory has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/community-content/CODEOWNERS) file under the `Community Content` section, pointing to the author or the author's team.
- [ ] Passes all the required formatting and linting checks. You can locally test with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).

### CODE REVIEW

File: .cloud-build/tests/select_notebooks_test.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+sys.path.append("..")
+
+from execute_changed_notebooks_helper import (load_results, select_notebook)
+
+bucket:str = "cloud-build-notebooks-presubmit"
+bucket_file: str = "build_results"
+
+accum = load_results(bucket, bucket_file)
+
+print(accum)
```
Comment (ivanmkc) [inline .cloud-build/tests/select_notebooks_test.py:11]: You can use pytest and assertions to make it a proper test.

Comment (andrewferlitsch) [reply .cloud-build/tests/select_notebooks_test.py:11]: Yes, that was my next todo to use pytest and then add test for the function select_notebook()

File: .cloud-build/execute_changed_notebooks_helper.py [hunk 2]
```diff
@@ -44,6 +44,8 @@
 WORKER_TIMEOUT_BUFFER_IN_SECONDS: int = 60 * 60
 PYTHON_VERSION = "3.9"  # Set default python version
 
+MAX_RESULTS_AGE: int = (60 * 60) * 24 * 60  # 60 days
```
[the remark this answers is not in the document]
Comment (ivanmkc) [reply .cloud-build/execute_changed_notebooks_helper.py:47]: Add units in name. i.e. MAX_RESULTS_AGE_SECONDS

Naming is a little ambiguous. What is the meaning of result?

File: .cloud-build/tests/select_notebooks_test.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    n_notselect = 0
+    for notebook in accum:
+        if select_notebook(notebook, accum, 50):
+            n_select += 1
+        else:
+            n_notselect += 1
+
+    print(f"SELECTED {n_select}, NOT SELECTED {n_notselect}")
+
+    assert n_select > 0
```
Comment (ivanmkc) [inline .cloud-build/tests/select_notebooks_test.py:32]: The method being tested is stochastic, so you probably want some sort of probabilistic test with sufficient samples to reduce false test errors.

Test for various combinations of `select_notebook` parameters, such as 0, 1, and in between.

File: .cloud-build/tests/select_notebooks_test.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import sys
+sys.path.append("..")
```
Comment (ivanmkc) [inline .cloud-build/tests/select_notebooks_test.py:2]: Oh don't do this. You can set PYTHONPATH in the envvar of the test runner instead.

File: .cloud-build/tests/select_notebooks_test.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+from execute_changed_notebooks_helper import (load_results, select_notebook)
+
+bucket:str = "cloud-build-notebooks-presubmit"
```
Comment (ivanmkc) [inline .cloud-build/tests/select_notebooks_test.py:6]: `bucket: str` <- add spacing.

File: .cloud-build/tests/select_notebooks_test.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+bucket_file: str = "build_results"
+accum = {}
```
Comment (ivanmkc) [inline .cloud-build/tests/select_notebooks_test.py:8]: Unit tests are supposed to be independent, so don't include global mutable state.

File: .cloud-build/tests/select_notebooks_test.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+def test_load_results():
+    global bucket, bucket_file, accum
```
Comment (ivanmkc) [inline .cloud-build/tests/select_notebooks_test.py:11]: Unit tests are supposed to be independent, so don't include global mutable state.

File: .cloud-build/tests/select_notebooks_test.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+    accum = load_results(bucket, bucket_file)
+
+    print(accum)
+
+    assert accum != {}
+
+def test_select_notebook():
+    global accum
```
Comment (ivanmkc) [inline .cloud-build/tests/select_notebooks_test.py:20]: Unit tests are supposed to be independent, so don't include global mutable state.

File: .cloud-build/tests/select_notebooks_test.py (hunk 1 above)
Comment (andrewferlitsch) [reply .cloud-build/tests/select_notebooks_test.py]: done

File: .cloud-build/tests/select_notebooks_test.py (hunk 5 above)
Comment (andrewferlitsch) [reply .cloud-build/tests/select_notebooks_test.py:6]: done

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewferlitsch) [reply .cloud-build/execute_changed_notebooks_helper.py:47]: done

File: .cloud-build/tests/select_notebooks_test.py (hunk 6 above)
Comment (andrewferlitsch) [reply .cloud-build/tests/select_notebooks_test.py:8]: done

File: .cloud-build/tests/select_notebooks_test.py (hunk 7 above)
Comment (andrewferlitsch) [reply .cloud-build/tests/select_notebooks_test.py:11]: done

File: .cloud-build/tests/select_notebooks_test.py (hunk 8 above)
Comment (andrewferlitsch) [reply .cloud-build/tests/select_notebooks_test.py:20]: done

File: .cloud-build/tests/select_notebooks_test.py (hunk 4 above)
Comment (andrewferlitsch) [reply .cloud-build/tests/select_notebooks_test.py]: removed

File: .cloud-build/tests/select_notebooks_test.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+def test_load_results():
+    bucket: str = "cloud-build-notebooks-presubmit"
+    bucket_file: str = "build_results"
+
+    accum = load_results(bucket, bucket_file)
+
+    print(accum)
+
+    assert accum != {}
```
Comment (ivanmkc) [inline .cloud-build/tests/select_notebooks_test.py:14]: assert len(accum) > 0

Comment (andrewferlitsch) [reply .cloud-build/tests/select_notebooks_test.py]: done

File: .cloud-build/tests/select_notebooks_test.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
+import sys
```
Comment (ivanmkc) [inline .cloud-build/tests/select_notebooks_test.py:1]: Remove unused import.

File: .cloud-build/tests/select_notebooks_test.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    bucket_file: str = "build_results"
+
+    accum = load_results(bucket, bucket_file)
+
+    print(accum)
+
+    assert len(accum) > 0
+
+def test_select_notebook():
+    bucket: str = "cloud-build-notebooks-presubmit"
```
Comment (ivanmkc) [inline .cloud-build/tests/select_notebooks_test.py:17]: nit: you can remove type-hint as that is inferred

**Review by ivanmkc — APPROVED** [review]
LGTM

**Pull request merged**